src/perf/tracepoint/format.cpp

```diff
@@ -30,34 +30,57 @@
 #include <boost/filesystem.hpp>
 #include <boost/filesystem/fstream.hpp>
 
+#include <cerrno>
+
 namespace lo2s
 {
 namespace perf
 {
 namespace tracepoint
 {
+EventFormat::ParseError::ParseError(const std::string& what, int error_code)
+: std::runtime_error{ what + ": " + std::strerror(error_code) }
+{
+}
 
 EventFormat::EventFormat(const std::string& name) : name_(name)
 {
+    using namespace std::string_literals;
+
     // allow perf-like name format which uses ':' as a separator
     std::replace(name_.begin(), name_.end(), ':', '/');
 
     boost::filesystem::path path_event = base_path_ / name_;
     boost::filesystem::ifstream ifs_id, ifs_format;
 
-    ifs_id.exceptions(std::ios::failbit | std::ios::badbit);
-    ifs_format.exceptions(std::ios::failbit | std::ios::badbit);
+    auto id_path = path_event / "id";
+    auto format_path = path_event / "format";
 
-    ifs_id.open(path_event / "id");
+    ifs_id.open(id_path);
     ifs_id >> id_;
 
-    ifs_format.open(path_event / "format");
+    if (ifs_id.fail())
+    {
+        throw ParseError{ "Failed to read tracepoint ID file "s + id_path.string(), errno };
```

Comment:
This whole checking is a) not symmetrical for both files and b) looks repetitive. You could refactor it to something like this:

```
read_file_checked(id_path, [this](auto& fs){
    fs >> this->id_;
});

read_file_checked(format_path, [this](auto& fs) {
    std::string line;
    while (std::getline(fs, line)) {
          this->parse_format_line(line);
    }
});
```

Reply:
Okay, disregard that, iostreams are broken.
*INSERT YET ANOTHER RANT*